app/ledger/neonLedger.js

```diff
@@ -0,0 +1,199 @@
+  async close () {
+    return this.device.close()
+  }
+
+  /**
+   * Retrieves the public key of an account from the Ledger.
+   * @param {number} [acct] - Account that you want to retrieve the public key from.
+   * @return {string} Public Key (Unencoded)
+   */
+  async getPublicKey (acct: number = 0): Promise<string> {
+    const res = await this.send('80040000', BIP44(acct), VALID_STATUS)
+    return res.toString('hex').substring(0, 130)
+  }
+
+  /**
+   * Sends an message with params over to the Ledger.
+   * @param {string} params - params as a hexstring
+   * @param {string} msg - Message as a hexstring
+   * @param {number[]} statusList - Statuses to return
+   * @return {Promise<Buffer>} return value decoded to ASCII string
+   */
+  async send (params: string, msg: string, statusList: number[]): Promise<Buffer> {
+    if (params.length !== 8) throw new Error(`params requires 4 bytes`)
+    // $FlowFixMe
+    const [cla, ins, p1, p2] = params.match(/.{1,2}/g).map(i => parseInt(i, 16))
+    return this.device.send(cla, ins, p1, p2, Buffer.from(msg, 'hex'), statusList)
+  }
+
+  /**
+   * Gets the ECDH signature of the data from Ledger using acct
+   * @param {string} data
+   * @param {number} [acct]
+   * @return {Promise<string>}
+   */
+  async getSignature (data: string, acct: number = 0): Promise<string> {
+    data += BIP44(acct)
+    let response = null
+    const chunks = data.match(/.{1,510}/g) || []
+    if (!chunks.length) throw new Error(`Invalid data provided: ${data}`)
+    for (let i = 0; i < chunks.length; i++) {
+      const p = i === chunks.length - 1 ? '80' : '00'
+      // $FlowFixMe
+      const chunk = chunks[i]
+      const params = `8002${p}00`
+      let [err, res] = await asyncWrap(this.send(params, chunk, VALID_STATUS))
+      if (err) {
+        const errCode = p === '00' ? '0' : '1'
+        console.log(`Signature Reponse An error occurred[${errCode}]:`, err)
+        throw new Error(`An error occurred[${errCode}]: ${err}`)
+      }
+      response = res
+    }
+    if (response === 0x9000) {
+      throw new Error(`No more data but Ledger did not return signature!`)
+    }
+    // $FlowFixMe
+    return assembleSignature(response.toString('hex'))
+  }
+}
+
+const assembleSignature = (response: string): string => {
+  let rLenHex = response.substring(6, 8)
+  let rLen = parseInt(rLenHex, 16) * 2
+  let rStart = 8
+  let rEnd = rStart + rLen
+
+  while ((response.substring(rStart, rStart + 2) === '00') && ((rEnd - rStart) > 64)) {
+    rStart += 2
+  }
+
+  let r = response.substring(rStart, rEnd)
+  let sLenHex = response.substring(rEnd + 2, rEnd + 4)
+  let sLen = parseInt(sLenHex, 16) * 2
+  let sStart = rEnd + 4
+  let sEnd = sStart + sLen
+
+  while ((response.substring(sStart, sStart + 2) === '00') && ((sEnd - sStart) > 64)) {
+    sStart += 2
+  }
+
+  let s = response.substring(sStart, sEnd)
+
+  while (r.length < 64) {
+    r = '00' + r
+  }
+
+  while (s.length < 64) {
+    s = '00' + s
+  }

```

Comment:
:neckbeard: You could use `String.prototype.padStart` in place of both these for-loops:

```js
r = r.padStart(64, '0')
s = s.padStart(64, '0')
```

Reply:
This is inherited code lol I will clean up the parsing tonight. I have a good idea of what is happening there.

Reply:
FWIW, I use the :neckbeard: emoji to represent "if you feel like it".  Don't feel obligated to refactor it.

Reply:
Cleaner code is always better :) Anyway looks like I would be adding on stuff if Coranos replies with the error code I should look out for.